gems/aws-partitions/lib/aws-partitions.rb

```diff
@@ -209,11 +209,27 @@ def default_partition_list
 
       # @return [Hash]
       # @api private
-      def defaults
-        @defaults ||= begin
-          path = File.expand_path('../../partitions.json', __FILE__)
-          defaults = JSON.load(File.read(path), freeze: true)
-          defaults.merge('partitions' => defaults['partitions'].dup)
+      json_freeze = begin
+        JSON.load('true', freeze: true)
+      rescue
```

Comment:
Apologies for additional churn. I'd rather not use exceptions for logic control flow - and also not use dynamic method definition..

You mentioned that JSON load takes an options hash even for early Ruby versions and it would ignore the `:freeze` option. The SDK at a minimum supports 1.9.3 and I see the options hash you are talking about here: https://ruby-doc.org/stdlib-1.9.3/libdoc/json/rdoc/JSON.html#method-i-load . I agree with Alex that the documentation isn't very clear here.. I guess the Ruby version wouldn't matter here, unless that Ruby version has an upper bound on the JSON version? My understanding is that it was built into stdlib but that could just be wrong. These errors were definitely strange as I agree with you that the option should have been ignored https://github.com/aws/aws-sdk-ruby/runs/1829320013.

Do you think a `JSON::VERSION` (or worst case a `RUBY_VERSION`) check would work instead? 

Perhaps the method would look something like:

```
 def defaults
   @defaults ||= begin
     path = File.expand_path('../../partitions.json', __FILE__)
     options = {}
     options[:freeze] = true if JSON::VERSION >= '2.4.0'
     defaults = JSON.load(File.read(path), options)
     defaults.merge('partitions' => defaults['partitions'].dup)
   end
 end
```

Or worst case, use `JSON.load(File.read(path), freeze: true)` when `JSON::VERSION >= '2.4.0'` and `JSON.load(File.read(path)` otherwise.

Reply:
> Do you think a JSON::VERSION (or worst case a RUBY_VERSION) check would work instead?

In my experience feature testing is better because it also handle future breakage. But sure no problem.

Reply:
I pushed a new commit that does a version check.

Reply:
Thanks! In the SDK we have many checks against RUBY_VERSION and some other version constants so we find it better to be consistent. I'm personally not a fan of rescue logic for control flow - it's a bit opinionated, but I get where you're coming from. In theory gems follow semantic versioning, so 2.4.0+ should have the feature, but then again, a gem could have a regression, so I see both points.